stdlib/SuiteSparse/src/cholmod.jl

```diff
@@ -228,116 +246,98 @@ struct C_Sparse{Tv<:VTypes} <: SuiteSparseStruct
     packed::Cint
 end
 
-# Corresponds to the exact definition of cholmod_sparse_struct in the library.
-# Useful when reading matrices of unknown type from files as in
-# cholmod_read_sparse
-struct C_SparseVoid <: SuiteSparseStruct
-    nrow::Csize_t
-    ncol::Csize_t
-    nzmax::Csize_t
-    p::Ptr{Cvoid}
-    i::Ptr{Cvoid}
-    nz::Ptr{Cvoid}
-    x::Ptr{Cvoid}
-    z::Ptr{Cvoid}
-    stype::Cint
-    itype::Cint
-    xtype::Cint
-    dtype::Cint
-    sorted::Cint
-    packed::Cint
-end
-
 mutable struct Sparse{Tv<:VTypes} <: AbstractSparseMatrix{Tv,SuiteSparse_long}
     ptr::Ptr{C_Sparse{Tv}}
     function Sparse{Tv}(ptr::Ptr{C_Sparse{Tv}}) where Tv<:VTypes
         if ptr == C_NULL
             throw(ArgumentError("sparse matrix construction failed for " *
                 "unknown reasons. Please submit a bug report."))
         end
+        s = unsafe_load(ptr)
+        if s.itype != ityp(SuiteSparse_long)
+            free!(ptr)
+            throw(CHOLMODException("itype=$(s.itype) not supported"))
+        elseif s.xtype != xtyp(Tv)
+            free!(ptr)
+            throw(CHOLMODException("xtype=$(s.xtype) not supported"))
+        elseif s.dtype != dtyp(Tv)
+            free!(ptr)
+            throw(CHOLMODException("dtype=$(s.dtype) not supported"))
+        end
         A = new(ptr)
         finalizer(free!, A)
         return A
     end
 end
 Sparse(p::Ptr{C_Sparse{Tv}}) where {Tv<:VTypes} = Sparse{Tv}(p)
 
+# Useful when reading in files, but not type stable
+function Sparse(p::Ptr{C_Sparse{Cvoid}})
+    if p == C_NULL
+        throw(ArgumentError("sparse matrix construction failed for " *
+                            "unknown reasons. Please submit a bug report."))
+    end
+    s = unsafe_load(p)
+    Tv = s.xtype == REAL ? Float64 : Complex{Float64}
+    Sparse(convert(Ptr{C_Sparse{Tv}}, p))
+end
+
 Base.unsafe_convert(::Type{Ptr{Tv}}, A::Sparse{Tv}) where {Tv} = getfield(A, :ptr)
 
 # Factor
+# Cvoid is used for pattern-only factors
+struct C_Factor{Tv<:VTypes} <: SuiteSparseStruct
+    n::Csize_t
+    minor::Csize_t
+    Perm::Ptr{SuiteSparse_long}
+    ColCount::Ptr{SuiteSparse_long}
+    @static if build_version >= v"2.1.0"
```

Comment:
Nice. I didn't realize you could do this.

Reply:
yeah, we do the same thing in LibGit2